lib/AST/Decl.cpp

```diff
@@ -5703,12 +5703,19 @@ void VarDecl::emitLetToVarNoteIfSimple(DeclContext *UseDC) const {
     if (FD && !FD->isMutating() && !FD->isImplicit() && FD->isInstanceMember()&&
         !FD->getDeclContext()->getDeclaredInterfaceType()
                  ->hasReferenceSemantics()) {
-      // Do not suggest the fix it in implicit getters
+      // Do not suggest the fix-it in implicit getters
       if (auto AD = dyn_cast<AccessorDecl>(FD)) {
         if (AD->isGetter() && !AD->getAccessorKeywordLoc().isValid())
           return;
+
+        // Do not suggest the fix-it if `Self` is a class type.
+        if (AD->getDeclContext()
+                ->getSelfTypeInContext()
+                ->isAnyClassReferenceType()) {
```

Comment:
In retrospect, should this just be `!doesContextHaveValueSemantics(AD->getDeclContext())` instead of what I said? (And whoops on `getSelfInterfaceType` vs. `getSelfTypeInContext`, yours is correct.)

Reply:
Hmm `doesContextHaveValueSemantics` is not accessible here though and it also calls `isReferenceTypeImpl()`, which is what `isAnyClassReferenceType()` and `hasReferenceSemantics()` calls (difference between the two that the former doesn't include functions but the latter does).

I think we don't need the `doesContextHaveValueSemantics` check as `getSelfTypeInContext()` will return the mapped `Self` type of the protocol or extension (it would be a `PrimaryArchetypeType`) and then the `isReferenceTypeImpl()` check would look into it to check if it requires class.

What do you think?

Reply:
Maybe we can move `doesContextHaveValueSemantics()` into `DeclContext`.

Reply:
I think `getSelfTypeInContext()` is going to be more expensive than checking against the generic signature (because the former would actually have to form the type), but either way is okay, as is moving the predicate into DeclContext. @slavapestov, any opinions?

Reply:
I've moved `doesContextHaveValueSemantics()` into `DeclContext` and switched to `getSelfTypeInContext()`, which is simpler to use, but I can revert back to the generic signature check if it's less expensive.